test/test_binary_ufuncs.py

```diff
@@ -573,56 +573,45 @@ def _do_pow_for_exponents(self, m1, exponents, pow_fn, atol):
                 self.assertEqual(res1, res2)
                 self.assertEqual(res1.dtype, expected_dtype)
 
-    def test_pow(self, device):
+    # not testing complex types due to https://github.com/pytorch/pytorch/pull/36793
+    @dtypes(*torch.testing.get_all_dtypes(include_bool=False, include_complex=False))
+    def test_pow(self, device, dtype):
```

Comment:
Let's talk more about the structure of this test now that the CPU internals are looking pretty good.

If I was going to test pow today I'd want to do two things:

- use numpy.power as a reference (https://numpy.org/doc/stable/reference/generated/numpy.power.html)
- use make_tensor() to produce a variety of tensors

In particular, I think you can simplify this test while extending its coverage by generating tensors using make_tensor(). You can generate empty tensors, some tensors with only a few elements (like 2), tensors with many elements (like 2D tensors with shape 513x513), and noncontiguous versions of all these tensors. make_tensor() also lets you specify a range for the tensors, so to avoid a base of zero, for example, you could generate strictly positive tensors and strictly negative tensors. Then this can just pass each of these pairs of tensors to pow() and compare the output with numpy.power. 

The one caveat with this approach is that numpy doesn't support bfloat16. For that particular case this test can compute in float32 on NumPy.

Does that make sense? Let me know. If it's confusing I can provide a code example.

cc @mattip who's looking at improving our architecture for binary tests.

Reply:
There are some other tests in `test_binary_ufuncs.py` that compare the output of the `pow` operator with that of `numpy.power`. I'll add more tensors to them, add more tests if required & will enable support for all supported dtypes.

Reply:
Hello @mruberry & @heitorschueroff, as `_test_pow` already compares the output of the `pow` operator with `numpy.power`, please confirm if a test such as the following one can replace` test_int_pow` for scalar exponents. Based on your feedback, I'd add tests for tensor exponents in the same method, and will refactor/add similar tests for other dtypes.

```
    def test_int_pow(self, device):

        def _test_integral_pow(dt, low, high, dev):
            test_cases = (
                ((3, 3), 4),
                ((2,), 4),
                ((1,), 2),
                ((513, 513), 4),
                ((5, 5, 5), 5),
                ((), 3),
            )
            for base, exp in test_cases:
                base_tensor = make_tensor(base, dtype=dt, device=dev, low=low, high=high)
                self._test_pow(base_tensor, exp)
                # test non-contiguous tensors as well
                if base_tensor.ndim <= 2:
                    self._test_pow(base_tensor.t(), exp)

        _test_integral_pow(torch.int8, -3, 4, device)
        _test_integral_pow(torch.uint8, 0, 4, device)
        _test_integral_pow(torch.int16, -5, 5, device)
        _test_integral_pow(torch.int64, -10, 10, device)
        _test_integral_pow(torch.int32, -10, 10, device)
```

Reply:
Oh, I see now that I can create discontiguous tensors with `make_tensor` as well.
So, I'll test every aspect that mruberry recommended.

Reply:
Hello @mruberry, thanks for your guidance!

I just pushed some changes in `test/test_binary_ufuncs.py`, as an example. In particular, I refactored `test_int_pow` to test more comprehensively. This test uses integral base & exponents & compares output with `numpy.power`, but I added more types of test inputs. Please confirm if [the changes](https://github.com/pytorch/pytorch/commit/d21da15c4612a33dcb0b8349cf56ba80b34ddc8f#diff-c99ae5b91314e5551d6b61d1d91539c550b844cfe38a7a063c44ae4d2b479c15) are in line with what you had [alluded to earlier](https://github.com/pytorch/pytorch/pull/50999/files#r593078503). If the changes look good, then I'll refactor some other tests & add more for other dtypes. Thank you!